Update DB.lua

JorMox opened the pull request:
db:_migrate - change to avoid error when PRAGMA call results in nil rather than number result
db:_drop - changed declaration into method rather than function form

vadi2 commented:
Tagging @Mudlet/lua-interface for review

Kind of want to see what happened to busted unit tests so we can add this to them as well, so this'll take a bit of time to process @JorMox

vadi2 commented:
So this is something cool - if the automated infrastructure builds worked right, https://transfer.sh/drnXK/Mudlet-3.0.0-testing-PR845-2455efe.dmg is an autogenerated .dmg with your improvements. Want to double-check that the fixes work @JorMox ?

JorMox commented:
Well, the changes definitely have been made in the auto-generated build, and everything appears to be working. Granted, I don't actually know what most of this stuff does, so me saying it is working is basically just saying it isn't generating errors, and I am able to successfully create a new database, add to an existing database schema, and properly store and retrieve information from such a database in the same way as I was able to before.

To put things in context, the _drop method is only ever user called, so the change should make no impact in a general sense, as it isn't even documented anywhere, so the chances of it being used are slim.

The _migrate method is only called by db:create, and after trying a wide variety of things to reproduce the error I was seeing (with the cur variable being nil), I determined that it only occurs when the database exists, is then deleted (in the OS, not the declaratory code in Mudlet), and then db:create is called again for that database. The proposed change to the if statement results in the problem being skipped over, but the database is not properly recreated, so I think the better approach would be to modify db:create in such a way that it tests to see if the file exists in reality. I think this would probably fit in line 438 where it is checking the db.__conn table for the existence of the database, presumably by way of io.exists(getMudletHomeDir() .. "/Database_" .. db_name .. ".db")

Thoughts?

JorMox commented:
Side note, it appears that we are using LuaSQL v. 2.3.0, and the most recent version is 2.3.5, released just this past February. We should look into updating our LuaSQL to see if that fixes some of the current issues.

vadi2 commented:
Making things more resilient in general is a good idea, but I'll need to
wrap my head around the DB functions again in order to comment
constructively. I suspect there might be a reason as to why underscored
functions weren't documented.


On Thu, 6 Apr 2017 4:12 pm JorMox, <notifications@github.com> wrote:

Well, the changes definitely have been made in the auto-generated build,
and everything appears to be working. Granted, I don't actually know what
most of this stuff does, so me saying it is working is basically just
saying it isn't generating errors, and I am able to successfully create a
new database, add to an existing database schema, and properly store and
retrieve information from such a database in the same way as I was able to
before.

To put things in context, the _drop method is only ever user called, so the
change should make no impact in a general sense, as it isn't even
documented anywhere, so the chances of it being used are slim.

The _migrate method is only called by db:create, and after trying a wide
variety of things to reproduce the error I was seeing (with the cur
variable being nil), I determined that it only occurs when the database
exists, is then deleted (in the OS, not the declaratory code in Mudlet),
and then db:create is called again for that database. The proposed change
to the if statement results in the problem being skipped over, but the
database is not properly recreated, so I think the better approach would be
to modify db:create in such a way that it tests to see if the file exists
in reality. I think this would probably fit in line 438 where it is
checking the db._*conn table for the existence of the database, presumably
by way of io.exists(getMudletHomeDir() .. "/Database*" .. db_name .. ".db")

Thoughts?

—
You are receiving this because you are on a team that was mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/Mudlet/Mudlet/pull/845#issuecomment-292186821>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAGxjH5x7XMepNNKRsv74egJ-U3CTqVMks5rtPK7gaJpZM4M1JDA>
.

JorMox commented:
I updated the if statement in db._migrate to check if type(cur) == "userdata"  (which matches a similar if statement later in the function), and added the proposed additional check in db:create, or not io.exists(getMudletHomeDir() .. "/Database_" .. db_name .. ".db")), so that if the database file no longer exists, a new connection will be made.

No idea how to detect a locked database file, or how to work around such a lock so that the database can be used.

SlySven commented:
Just as well as it is tagged for the lua-interface team - I'm not up to reviewing this... :disappointed:

vadi2 commented:
@keneanung?

vadi2 commented:
Thanks for the improvements @JorMox :)

Got anything else, send another PR in!
